File: client/go/src/example/card/card.go
package card

import (
	"fmt"
	"context"
	"biostar/service/card"
	"google.golang.org/grpc"
)

type CardSvc struct {
	client card.CardClient
}

func NewCardSvc(conn *grpc.ClientConn) *CardSvc {
	return &CardSvc{
		client: card.NewCardClient(conn),
	}
}

func (s *CardSvc) Scan(deviceID uint32) (*card.CardData, error) {
	req := &card.ScanRequest{
		DeviceID: deviceID,
	}

	resp, err := s.client.Scan(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot scan a card: %v\n", err)

		return nil, err
	}

	return resp.GetCardData(), nil
}

func (s *CardSvc) GetBlacklist(deviceID uint32) ([]*card.BlacklistItem, error) {
	req := &card.GetBlacklistRequest{
		DeviceID: deviceID,
	}

	resp, err := s.client.GetBlacklist(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot get blacklist: %v\n", err)

		return nil, err
	}

	return resp.GetBlacklist(), nil
}

func (s *CardSvc) AddBlacklist(deviceID uint32, cardInfos []*card.BlacklistItem) error {
	req := &card.AddBlacklistRequest{
		DeviceID: deviceID,
		CardInfos: cardInfos,
	}

	_, err := s.client.AddBlacklist(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot add blacklist items: %v\n", err)

		return err
	}

	return nil
}


func (s *CardSvc) DeleteBlacklist(deviceID uint32, cardInfos []*card.BlacklistItem) error {
	req := &card.DeleteBlacklistRequest{
		DeviceID: deviceID,
		CardInfos: cardInfos,
	}

	_, err := s.client.DeleteBlacklist(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot delete blacklist items: %v\n", err)

		return err
	}

	return nil
}

File: client/python/example/card/card.py
import grpc

import card_pb2_grpc
import card_pb2


class CardSvc:
  stub = None

  def __init__(self, channel): 
    try:
      self.stub = card_pb2_grpc.CardStub(channel)
    except grpc.RpcError as e:
      print(f'Cannot get the card stub: {e}')
      raise

  def scan(self, deviceID):
    try:
      response = self.stub.Scan(card_pb2.ScanRequest(deviceID=deviceID))
      return response.cardData
    except grpc.RpcError as e:
      print(f'Cannot scan a card: {e}')
      raise

  def getBlacklist(self, deviceID):
    try:
      response = self.stub.GetBlacklist(card_pb2.GetBlacklistRequest(deviceID=deviceID))
      return response.blacklist
    except grpc.RpcError as e:
      print(f'Cannot get the blacklist: {e}')
      raise

  def addBlacklist(self, deviceID, cardInfos):
    try:
      self.stub.AddBlacklist(card_pb2.AddBlacklistRequest(deviceID=deviceID, cardInfos=cardInfos))
    except grpc.RpcError as e:
      print(f'Cannot add the cards to the blacklist: {e}')
      raise

  def deleteBlacklist(self, deviceID, cardInfos):
    try:
      self.stub.DeleteBlacklist(card_pb2.DeleteBlacklistRequest(deviceID=deviceID, cardInfos=cardInfos))
    except grpc.RpcError as e:
      print(f'Cannot delete the cards from the blacklist: {e}')
      raise

      
